tests/e2e/stateapp/stateapp_test.go

```diff
@@ -0,0 +1,297 @@
+// +build e2e
+
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package stateapp_e2e
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"reflect"
+	"testing"
+
+	"github.com/dapr/dapr/tests/e2e/utils"
+	kube "github.com/dapr/dapr/tests/platforms/kubernetes"
+	"github.com/dapr/dapr/tests/runner"
+	guuid "github.com/google/uuid"
+	"github.com/stretchr/testify/require"
+)
+
+const appName = "stateapp"     // App name in Dapr.
+const healthChecks = 3         // Number of get calls before starting tests.
+const maxEntries = 10          // Some reasonable number.
+const testManyEntriesCount = 5 // Anything between 1 and the number above (inclusive).
+
+type testCommandRequest struct {
+	RemoteApp string `json:"remoteApp,omitempty"`
+	Method    string `json:"method,omitempty"`
+}
+
+// appState represents a state in this app.
+type appState struct {
+	Data string `json:"data,omitempty"`
+}
+
+// daprState represents a state in Dapr.
+type daprState struct {
+	Key   string    `json:"key,omitempty"`
+	Value *appState `json:"value,omitempty"`
+}
+
+// requestResponse represents a request or response for the APIs in the app.
+type requestResponse struct {
+	States []daprState `json:"states,omitempty"`
+}
+
+// represents each step in a test since it can make multiple calls.
+type testStep struct {
+	command          string
+	request          requestResponse
+	expectedResponse requestResponse
+}
+
+type testCase struct {
+	name  string
+	steps []testStep
+}
+
+// internal key value structure to simplify code given bloated app API.
+type simpleKeyValue struct {
+	key   string
+	value *string
+}
+
+// creates a requestResponse based on an array of key value pairs.
+func newRequestResponse(keyValues ...simpleKeyValue) requestResponse {
+	var daprStates = make([]daprState, 0, len(keyValues))
+	for _, keyValue := range keyValues {
+		if keyValue.value != nil {
+			daprStates = append(daprStates, daprState{
+				keyValue.key,
+				&appState{
+					*keyValue.value,
+				},
+			})
+		} else {
+			daprStates = append(daprStates, daprState{
+				keyValue.key,
+				nil,
```

Comment:
Re: comment above, this doesn't have to be `nil` if the value isn't a pointer to a string. The check for the value, since it is a string, can be made using `!= ""`.

Reply:
I thought about that but wanted to keep empty string as a valid value.

Reply:
Fixed by using interface to keep the nil check.